go/cmd/dolt/commands/mysql_rows_iter.go

```diff
@@ -0,0 +1,278 @@
+package commands
+
+import (
+	dsql "database/sql"
+	"fmt"
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/parse"
+	"io"
+	"reflect"
+	"strings"
+)
+
+// MySqlRowsIter is an iterator over rows returned by MySQL
+type MySqlRowsIter struct {
+	rows        *dsql.Rows
+	types       []sql.Type
+	columnTypes []*dsql.ColumnType
+	schema      sql.Schema
+}
+
+var _ sql.RowIter = (*MySqlRowsIter)(nil)
+
+// NewMySqlRowsIter creates a new MySqlRowsIter
+func NewMySqlRowsIter(ctx *sql.Context, rows *dsql.Rows) (*MySqlRowsIter, error) {
+	cols, err := rows.ColumnTypes()
+	if err != nil {
+		return nil, err
+	}
+	types, err := schemaToFields(ctx, cols)
+	if err != nil {
+		return nil, err
+	}
+	sch := make(sql.Schema, len(cols))
+	for i, col := range cols {
+		isNullable, ok := col.Nullable()
+		if !ok {
+			isNullable = true
+		}
+		sch[i] = &sql.Column{
+			Name:     col.Name(),
+			Type:     types[i],
+			Nullable: isNullable,
+		}
+	}
+
+	return &MySqlRowsIter{
+		rows:        rows,
+		types:       types,
+		schema:      sch,
+		columnTypes: cols,
+	}, nil
+}
+
+// Next returns the next sql.Row until all rows are returned at which point (nil, io.EOF) is returned.
+func (b MySqlRowsIter) Next(_ *sql.Context) (sql.Row, error) {
+	rows := b.rows
+	if rows.Next() {
+		row, err := scanResultRow(rows, b.columnTypes)
+		return row, err
+	}
+	return nil, io.EOF
+}
+
+// Close closes the iterator.
+func (b MySqlRowsIter) Close(_ *sql.Context) error {
+	return b.rows.Close()
+}
+
+// Schema returns the schema for the rows returned by this iterator.
+func (b MySqlRowsIter) Schema() sql.Schema {
+	return b.schema
+}
+
+func scanResultRow(results *dsql.Rows, cols []*dsql.ColumnType) (sql.Row, error) {
```

Comment:
You're going to need a lot of tests for this to convince me it's right. This feels like a lot of code what we could get wrong.